### ISSUE TITLE
iOS11 Freezing during P2P connection with Chrome / FF

### ISSUE THREAD

**Issue by MilllerTime** [twilio/twilio-video.js#192]
- [X] I have verified that the issue occurs with the latest twilio-video.js release and is not marked as a known issue in the [CHANGELOG.md](https://github.com/twilio/twilio-video.js/blob/master/CHANGELOG.md).
 - [X] I reviewed the [Common Issues](https://github.com/twilio/twilio-video.js/blob/master/COMMON_ISSUES.md) and open GitHub issues and verified that this report represents a potentially new issue.
 - [x] I verified that the [Quickstart](https://github.com/twilio/video-quickstart-js) application works in my environment.

**Code to reproduce the issue:**

I hoped to be able to repro using only the video-quickstart project, but I'm unable to access it on an iOS device since the local Express server isn't configured to serve over HTTPS, a requirement for Safari. I can allow WebRTC on insecure origins using Safari 11 on High Sierra, but apparently not on iOS. Also, am I correct in assuming the quickstart is using a group room? I realize that is currently experimental with Safari 11, but I'm experiencing crashes using a peer-to-peer room, browser-to-browser, which I understand should work. Still, if there is a method of running the quickstart app on an iOS device that I haven't done, I can certainly try it.

**Expected behavior:**

Currently as long as an iOS 11 device is not involved in a P2P room, both parties receive audio/video and everything is great. This is the expected behavior, and leads me to believe my code isn't an issue. Also per Twilio's release notes, starting with 1.2.1, there is experimental support for Safari 11 and:

> P2P Rooms with browser-based Participants should work.

This is my use case and the expected result.

**Actual behavior:**

When Safari on an iOS 11 device (have tested both iPhone 6 and iPhone 7 Plus) tries to join a P2P room with Chrome or FF on the other end, the connection seems to work at first, but then iOS freezes within a few seconds. The video stream from the iOS device is still live in the other browser for a few more seconds, but it then freezes as well. At that point, the iOS device becomes totally unresponsive. The home button fails to work. Sometimes the power button is unresponsive as well. Other times it turns the screen off, but then the phone fails to wake back up. Either way, the device is essentially bricked until a hard restart is performed. I have tested with both iOS 11.0.2 and the latest 11.0.3 with the same results.

It's worth noting that making a P2P connection from Safari 11 on iOS to Safari 11 on Mac OS High Sierra works flawlessly. Figures 🤷‍♂️ 

This may be related to #189, but rotating / zooming is definitely not necessary to trigger the freeze.

I just captured a log from an iPhone 6 running iOS 11.0.3 when this freeze occurred. The other client was Chrome 61 on Windows 10. The connection was fine for a few seconds, then the UI on the iPhone froze, then became responsive again briefly before finally freezing completely and bricking the device. Interestingly, I continued to see streaming video from the phone on Chrome, but it was delayed by about 15 seconds (typically this freezes too, as mentioned above). Twilio also continued to print new logs while the phone was frozen. After several minutes of continued unresponsiveness, I saved the following log and hard rebooted the phone.

[iOS-Twilio-log.txt](https://github.com/twilio/twilio-video.js/files/1380564/iOS-Twilio-log.txt)

**Software versions:**

**Browser(s):** Safari 11 => Firefox 56 & Chrome 61
**Operating System:** iOS 11.0.2 / 11.0.3 => Windows 10 & Mac OS High Sierra
**twilio-video.js:** 1.5.0
**Third-party libraries:** React

**Comment by acoolprogrammer123** [comment]
Hi from https://github.com/twilio/twilio-video.js/issues/189, can confirm that I have seen similar issues without rotating or zooming given a short bit of time.  I am running the quick start demo for my back end.

**Comment by MilllerTime** [comment]
@acoolprogrammer123 Hi 😄 In your issue you stated that things run well for a couple minutes so long as you don't interfere. I'm not getting anywhere near that amount of stability, instead I'm usually seeing the freeze occur within mere seconds, and always under 20 sec. Do you _reliably_ get a few minutes of connection time?

It does _seem_ like we're hitting the same issue, I just wanted to make it clear that this is occurring in a P2P room and include my system details.

> I am running the quick start demo for my back end.

Are you saying you built a custom front end on top of the quick start demo?

**Comment by rfbrazier** [comment]
Hello all,

Thanks for this report--we're actively looking into it. 

> When Safari on an iOS 11 device (have tested both iPhone 6 and iPhone 7 Plus) tries to join a P2P room with Chrome or FF on the other end, the connection seems to work at first, but then iOS freezes within a few seconds.

We're trying to reproduce this issue, but aren't getting the same behavior. We've been able to reproduce a freeze in iOS Safari 11 <-> Firefox (which we believe we'll have corrected in the next release), but cannot reproduce a freeze in iOS Safari 11<->Chrome. 

A couple questions:

* Does the freeze happen on every call, or just intermittently?
* Do you have an instance of your app we can access to test and attempt to reproduce?

**Comment by acoolprogrammer123** [comment]
Chiming in...  I am 3 for 3 in getting an instant freeze on the iOS side... but only after setting my Twilio Console to Group Rooms, and then back to Peer to Peer, saving and then connecting to the quick start demo.  Not sure what the propagation time time is on that change, I am waiting ~15 seconds.

Other than that freezes are only intermittent (unless I rotate the phone a few times, then it is always almost instantly).  Only way recover from these freezes is a hard restart of the phone.

This is running the quickStart code on IIS 7.5, iPhone 6 on iOS 11.0.3 & Safari, Win 7 Chrome 61.0.3163.100 TwilioJS 1.5.0, with the only change to the quickstart being loading the js library via <script src="//media.twiliocdn.com/sdk/js/video/v1/twilio-video.min.js"></script>.  So no, no custom front end for this test, just the quickStart code.

**Comment by MilllerTime** [comment]
@rfbrazier Thanks for following up!

Definitely looking forward to trying the Firefox fix on this end.

> Does the freeze happen on every call, or just intermittently?

It's very consistent. Out of ~10 calls, only one did not totally brick the phone. However, it would still freeze for 15 - 30 seconds at a time, render a couple frames, freeze again, and so on.

> Do you have an instance of your app we can access to test and attempt to reproduce?

Unfortunately nothing publically accessible, and it would be non-trivial to spin up something for you to test. That said I'll see if we can host a reduced test case, perhaps even the quickstart code like @acoolprogrammer123 is doing. I'll get back to you about that on Monday.

Alex, thanks for the extra info!

Have a great weekend 🙂

**Comment by MilllerTime** [comment]
@rfbrazier We've installed the video-quickstart demo app on our IIS server (IIS 8.5) and ensured the created rooms are using P2P. We're still seeing consistent crashes on iOS when communicating with Chrome, so it does not appear to be an issue related to our application. As noted before, Chrome to Chrome still holds a connection just fine, as does Safari to Safari.

Just to be clear, I'm testing with Chrome in Incognito mode with all extensions disabled so nothing crazy should be happening there. Also, we're not doing anything fancy at all on the networking side.

Given that we're hitting this issue with the quickstart alone, is there anything else that would be helpful for us to try?

**Comment by markandrus** [comment]
@MilllerTime 

> I hoped to be able to repro using only the video-quickstart project, but I'm unable to access it on an iOS device since the local Express server isn't configured to serve over HTTPS, a requirement for Safari.

One trick I sometimes use is [ngrok](https://ngrok.com/)—it's free and lets you setup a temporary HTTPS proxy to your local server.

> Also, am I correct in assuming the quickstart is using a group room?

The QuickStart will use whatever default is configured for your account.

@acoolprogrammer123 

> Other than that freezes are only intermittent (unless I rotate the phone a few times, then it is always almost instantly). Only way recover from these freezes is a hard restart of the phone.

I'm trying this in a P2P demo app I created ([draw-with-twilio.appspot.com](https://draw-with-twilio.appspot.com)). I am doing testing iOS 11.0 (15a370a) with twilio-video.js@1.5.1 (includes performance fix for Firefox). Even with a lot of rotations, things are OK in both

* Chrome 61
* Firefox 56

I see that I have an iOS update available. I'll be updating my test device (iPhone 6s) and see if there's potentially a regression in newer iOS releases.

**Comment by markandrus** [comment]
I've tested again with an iPhone 6s and now iOS 11.1 (15B5086a). I am not able to reproduce with either Chrome or Safari. I do eventually see the video Safari is _receiving_ from Firefox freeze (although the received audio and data is fine), but I think this is an unrelated issue with Firefox.

**Comment by markandrus** [comment]
Sounds like @MilllerTime and @acoolprogrammer123 repro-ed with 11.0.2 and 11.0.3—i'll get off the iOS beta channel and try downgrading to either of these.

**Comment by MilllerTime** [comment]
@markandrus

I'll have to check out ngrok - looks like a pretty sweet tool! Also makes sense the quickstart uses our account default for room creation - thanks for clarifying.

> I'm trying this in a P2P demo app I created (draw-with-twilio.appspot.com).

Interesting, your custom demo works fine on this end with iOS <-> Chrome (both data and audio/video)

I just rebuilt both our own app as well as Twilio's quickstart demo to use twilio-video.js@1.5.1, and I'm still experiencing the same crash with both of those applications.  It's very intriguing that your draw-with-twilio demo works here, but the quickstart doesn't.

- Does the quickstart app work for you going from iOS <-> Chrome 61?
- Are there any notable differences between the quickstart and your drawing app?

**Comment by acoolprogrammer123** [comment]
I tried draw-with-twilio.appspot.com on my 6, and it survived more rotations, but still hung after about a minute of rotating the phone.  Averaged about 20ish rotations per try before the hang, but still saw the hang every single time, with one occurring after a single rotation.  

I also think this might be on a different version of the library because the hang didn't freeze the whole phone like my quickstart code does, just froze the remote participants video on both sides (local video was still updating).  I tried poking around and saw that draw-with-twilio.appspot.com was still referencing sip.js in index.js which I thought was factored out in 1.2.2?

Will try updating to the 1.1 iOS beta to see if I can reproduce your no-hangs if @markandrus will try a 40 rotation run on iOS 1.0.3 :)

**Comment by acoolprogrammer123** [comment]
This might be related to this group of issues (main thread links to a few others)

https://bugs.webkit.org/show_bug.cgi?id=176439

**Comment by MilllerTime** [comment]
@acoolprogrammer123 Well, that 100% describes the issue we're seeing. Nice find, and thanks for sharing the link!

From the comments on that bug, it unfortunately does not seem there is any viable workaround aside from ensuring that in a portrait orientation the remote video is displayed without cropping, either at the browser level or within an HTML element. That is not feasible for us. Will be keeping an eye on that issue.

**Comment by markandrus** [comment]
[Youenn mentions](https://bugs.webkit.org/show_bug.cgi?id=176439#c23) that the latest iOS 11.2 beta may fix this.

**Comment by markandrus** [comment]
I will close this issue here, since I think the issue stems from the one identified on https://github.com/twilio/twilio-video.js/issues/176. We can continue to track there.

**Comment by MilllerTime** [comment]
For the record, this issue no longer appears for us on iOS 11.2, as Youenn has suggested! 🎉
